synapse/storage/databases/main/event_federation.py

```diff
@@ -922,6 +962,26 @@ def _get_backfill_events(self, txn, room_id, event_list, limit):
                 if row[1] not in event_results:
                     queue.put((-row[0], row[1]))
 
+            # From each event, go forwards through successors to find chains of
+            # historical backfilled floating outliers
+            successorQueue = PriorityQueue()
+            successorQueue.put(event_id)
+            while not successorQueue.empty() and len(event_results) < limit:
+                try:
+                    event_id = successorQueue.get_nowait()
+                except Empty:
+                    break
+
+                successor_event_ids = self.get_successor_events_txn(txn, [event_id])
+                for successor_event_id in successor_event_ids:
+                    # Skip any branches we have already gone down
+                    if successor_event_id in event_results:
+                        continue
+
+                    event_results.add(successor_event_id)
+
+                    successorQueue.put(successor_event_id)
```

Comment:
I'm confused what this is doing? Shouldn't we be looking at the marker events here? I don't think we want to be looking at the prev events anymore?

Reply:
This is from the discussion in https://github.com/matrix-org/synapse/pull/9247#discussion_r622796142 before we decided on "marker" and "insertion" events. It goes down successor branches to also pull in the historical messages.

> > We'll probably want to extend the `_get_backfill_events` to also pull in these "new" historic events as well.
> 
> Thanks for the tip! I'll look into making it traverse the successors looking for `m.historical`

---

From our sync call, we mentioned [only putting the "marker" and "insertion" events in place with non of the backing logic](https://docs.google.com/document/d/1JJDuPfcPNX75fprdTWlxlaKjWOdbdJylbpZ03hzo638/edit#bookmark=id.4v1gkxb0eccu) so they are there and we can iterate on the logic behind them later. This way Beeper can use this now as we suss this out more.

Reply:
Ok, yup. Should we remove this code for now then?

Reply:
In terms of how we want to settle on navigating historical events:

Can we rely on the successor exploration as we iterate? It sounds like you weren't expecting it.

---

> Shouldn't we be looking at the marker events here? I don't think we want to be looking at the prev events anymore?

In order to make this work, should we add a new `insertion_event_lookups` table to store which events are marked as insertion points? It stores, `insertion_event_id` and  `insertion_prev_event_id`

This way when we're iterating back in time, we can check whether the current `event_id` is `insertion_prev_event_id` in the lookup table. When it matches, we can iterate from `insertion_event_id` by successors up each chunk. Are there performance problems with iterating successors?

![](https://user-images.githubusercontent.com/558581/118685772-fbc0f080-b7c8-11eb-96a1-793ef9222928.png)


---


When a remote federated homeserver, receives a "marker" event, it could

 - Backfill the "insertion" event immediately to add it to the `insertion_event_lookups` table. The rest of the chunk does not backfill yet.
 - Add the "insertion" event to the `event_backward_extremities`. When someone uses scrollback, it will request the historical messages and the navigation strategy above will return the correct messages.

Reply:
> Ok, yup. Should we remove this code for now then?

**Yes, we can delete it!** We better clarified this in a call this morning. I was confused about this piece being necessary for navigating the historical messages but it's just navigating them in the backfill response (as the file and function indicate 🤦)